### PR TITLE
Che bato/play from other zone

### PR DESCRIPTION
**Pull request by CheBato**
Added the functionality to play from other locations beside from hand.

- Added Ezra bridge

EzraBridger

- Added Brutal traditions (https://swudb.com/card/shd/038/brutal-traditions)

BrutalTraditions

### CODE REVIEW

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
+describe('Ezra Bridger', function() {
+    integration(function(contextRef) {
+        describe('Ezra Bridger\'s ability', function() {
```
[the remark this answers is not in the document]
Comment (adbronico) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:3]: Also a test for when Ezra's controller's deck is empty

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+describe('Brutal Traditions', function() {
+    integration(function(contextRef) {
+        describe('Brutal Tradition\'s ability', function() {
```
Comment (adbronico) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:3]: Can we add a test that ensures the Brutal Tradition can be played from discard multiple times per phase?

File: server/game/cards/01_SOR/units/EzraBridger.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                (context) => ({ target: context.source.controller.getTopCardOfDeck() })
+            ),
+            then: (thenContext) => {
+                const topCardOfDeck = thenContext.source.controller.getTopCardOfDeck();
+                return {
+                    title: 'Look at the top card of your deck.',
+                    thenCondition: (context) => context.source.controller.drawDeck.length > 0,
+                    targetResolver: {
+                        mode: TargetMode.Select,
+                        optional: true,
```
Comment (adbronico) [inline server/game/cards/01_SOR/units/EzraBridger.ts:30]: I think we can consider removing optional: true here. The card does say "may" but in this case Leave it on top of the deck serves the same purpose.

File: server/game/core/Player.js [hunk 4]
```diff
@@ -67,7 +67,9 @@ class Player extends GameObject {
 
         this.playableLocations = [
             new PlayableLocation(PlayType.PlayFromHand, this, Location.Hand),
-            new PlayableLocation(PlayType.Smuggle, this, Location.Resource)
+            new PlayableLocation(PlayType.Smuggle, this, Location.Resource),
+            new PlayableLocation(PlayType.PlayFromOutOfPlay, this, Location.Deck),
+            new PlayableLocation(PlayType.PlayFromOutOfPlay, this, Location.Discard),
```
Comment (adbronico) [inline server/game/core/Player.js:72]: @AMMayberry1 Do we want two records for one play type like this, or do we want to enhance the location to accept an array?

**Review by adbronico — DISMISSED** [review]
There's a few more test cases we should add. I'll let Veld review the AbilityContext changes since I'm not as confident in what should be changed there if anything.

File: server/game/cards/01_SOR/units/EzraBridger.ts (hunk 3 above)
Comment (CheBato) [reply server/game/cards/01_SOR/units/EzraBridger.ts:30]: Yes Sir i'll remove the extra optional

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                expect(context.player1).toHaveExactPromptButtons(['Play it', 'Discard it', 'Leave it on top of your deck']);
+
+                // check that the damage was done before player1 clicks prompt
+                expect(context.ezraBridger.damage).toBe(3);
+                expect(context.deathTrooper).toBeInLocation('discard');
+
+                // Leave it on top of the deck
+                const beforeActionDeck = context.player1.deck;
+                context.player1.clickPrompt('Leave it on top of your deck');
+                expect(context.player1.deck).toEqual(beforeActionDeck);
```
[the remark this answers is not in the document]
Comment (CheBato) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:48]: yes sir i updated the test cases

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                const reset = () => {
+                    context.ezraBridger.exhausted = false;
+                    context.player2.passAction();
+                };
+
+                // CASE 1: We do nothing and put the top card back
+                context.player1.clickCard(context.ezraBridger);
+                context.player1.clickCard(context.deathTrooper);
+
+                // TODO: we need a 'look at' prompt for secretly revealing, currently chat logs go to all players
```
[the remark this answers is not in the document]
Comment (CheBato) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:37]: Added the TODO

[the remark this answers is not in the document]
Comment (CheBato) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:37]: Added a TODO to the lookAtSystem!

File: server/game/gameSystems/PutIntoPlaySystem.ts [hunk 7]
```diff
@@ -56,11 +56,10 @@ export class PutIntoPlaySystem<TContext extends AbilityContext = AbilityContext>
     public override canAffect(card: Card, context: TContext): boolean {
         const contextCopy = context.copy({ source: card });
         const player = this.getPutIntoPlayPlayer(contextCopy);
-        const location = card.location;
-
+        const playAbleLocations = [Location.Resource, Location.Deck, Location.Discard];
         if (!super.canAffect(card, context)) {
             return false;
-        } else if (!card.canBeInPlay() || card.isInPlay() || (card.facedown && card.location !== Location.Resource)) {
+        } else if (!card.canBeInPlay() || card.isInPlay() || (card.facedown && !playAbleLocations.includes(card.location))) {
```
Comment (AMMayberry1) [inline server/game/gameSystems/PutIntoPlaySystem.ts:62]: Maybe we just remove the facedown check completely since I think it's actually impossible for it to fail now? The only facedown locations are resource and deck and those now have exceptions

File: server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import { StateWatcherName } from '../core/Constants';
+import { StateWatcherRegistrar } from '../core/stateWatcher/StateWatcherRegistrar';
+
+export interface CardDiscardedEntry {
+    card: Card;
+    controlledBy: Player;
+    discardedFromLocation: Location;
+}
+
+export type ICardsDiscardedThisPhase = CardDiscardedEntry[];
```
Comment (AMMayberry1) [inline server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts:15]: Let's move this file and the related changes out of this PR since it's for the Kylo PR and therefore isn't being tested. If we check it in then people will assume it's tested and working.

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                    player2: {
+                        base: 'dagobah-swamp',
+                        groundArena: ['death-trooper']
+                    }
+                });
+            });
+
+            it('should trigger when he completes an attack', function () {
+                const { context } = contextRef;
+                const reset = () => {
```
[the remark this answers is not in the document]
Comment (AMMayberry1) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:26]: I've been encouraging people to do them for now, it makes the long multi-case tests cleaner. We're going to have do a massive refactor anyway if we ever actually do get to the point of having a better test harness, and I don't think having or not having these will have a real impact on the amount of work.

If anything they would make it easier to refactor since we can tell right away which lines are just state reset and aren't part of a specific case.

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                expect(context.player1).toBeActivePlayer();
+                expect(context.player1).toBeAbleToSelect(context.brutalTraditions);
+
+                context.player1.clickCard(context.brutalTraditions);
+                expect(context.vigilantHonorGuards.upgrades).toEqual([context.brutalTraditions]);
+                expect(context.vigilantHonorGuards.getPower()).toBe(5);
+                expect(context.vigilantHonorGuards.getHp()).toBe(8);
+            });
+
+            it('should not be able to play it from the discard pile.', function () {
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:38]: Actually we can test two things at once here - if we merge this into the above test but do it after `moveToNextActionPhase()`, we confirm that the condition is over when the phase changes and that the card is not playable when the condition is not met

File: server/game/core/ability/PlayCardAction.ts [hunk 11]
```diff
@@ -59,7 +59,7 @@ export abstract class PlayCardAction extends PlayerAction {
             player: player,
             source: this.card,
             stage: Stage.PreTarget,
-            costAspects: this.card.aspects
+            costAspects: this.card.aspects,
```
[the remark this answers is not in the document]
Comment (CheBato) [reply server/game/core/ability/PlayCardAction.ts:62]: Yes sir my i apoligize.

[the remark this answers is not in the document]
Comment (CheBato) [reply server/game/core/ability/PlayCardAction.ts:62]: i think the , was leftover from when i attempted to add playType here to test it.

File: server/game/core/Player.js [hunk 12]
```diff
@@ -681,6 +683,10 @@ class Player extends GameObject {
                 aspects = smuggleInstance.aspects;
                 cost = smuggleInstance.cost;
                 break;
+            case PlayType.PlayFromOutOfPlay:
```
[the remark this answers is not in the document]
Comment (CheBato) [reply server/game/core/Player.js:686]: ooooh i didn't know you can do this thats pretty neet!

File: server/game/core/gameSteps/prompts/ResourcePrompt.ts [hunk 13]
```diff
@@ -67,7 +67,6 @@ export class ResourcePrompt extends AllPlayerPrompt {
     public override onCardClicked(player: Player, card: Card) {
         Contract.assertNotNullLike(player);
         Contract.assertNotNullLike(card);
-
```
[the remark this answers is not in the document]
Comment (CheBato) [reply server/game/core/gameSteps/prompts/ResourcePrompt.ts:70]: yessir i will go back to space being there

File: server/game/gameSystems/PutIntoPlaySystem.ts (hunk 7 above)
Comment (CheBato) [reply server/game/gameSystems/PutIntoPlaySystem.ts:62]: removed the facedown check!

Comment (CheBato) [reply server/game/gameSystems/PutIntoPlaySystem.ts:62]: Unfortunatelly by removing the check multiple tests failed like timely intervention so I'm assuming we need to keep it.

File: server/game/gameSystems/GameSystemLibrary.ts [hunk 14]
```diff
@@ -205,6 +205,9 @@ export function payResourceCost<TContext extends AbilityContext = AbilityContext
         )
     );
 }
+export function playCardFromOutOfPlay<TContext extends AbilityContext = AbilityContext>(propertyFactory: PropsFactory<Omit<IPlayCardProperties, 'optional'>, TContext> = {}): PlayCardSystem<TContext> {
+    return new PlayCardSystem(propertyFactory);
```
[the remark this answers is not in the document]
Comment (CheBato) [reply server/game/gameSystems/GameSystemLibrary.ts:209]: Set it to default seems like it works :)

File: server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts (hunk 8 above)
Comment (CheBato) [reply server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts:15]: Yes sir

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts (hunk 2 above)
Comment (CheBato) [reply test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:3]: added the test!

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts (hunk 10 above)
Comment (CheBato) [reply test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts]: yessir! Made them into one!

File: server/game/cards/01_SOR/units/EzraBridger.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+                onAttackCompleted: (event, context) => event.attack.attacker === context.source,
+            },
+            immediateEffect: AbilityHelper.immediateEffects.lookAt(
+                (context) => ({ target: context.source.controller.getTopCardOfDeck() })
+            ),
+            then: (thenContext) => {
+                const topCardOfDeck = thenContext.source.controller.getTopCardOfDeck();
+                return {
+                    title: 'You may play it, discard it, or leave it on top of your deck.',
+                    thenCondition: (context) => context.source.controller.drawDeck.length > 0,
```
Comment (AMMayberry1) [inline server/game/cards/01_SOR/units/EzraBridger.ts:27]: I've had a realization recently that cards of the form "look at the top card of your deck. do [X] with it" might actually be best represented with `ifYouDo` instead of `then` with a `thenCondition`. The card doesn't explicitly say "if you do" but mechanically it's understood that if you don't look at the top card for any reason (including the deck being empty) then the second part of the effect should fizzle.

Could you try making that change here? There is an option to have an `ifYouDoContext` but I think you won't need it since you're just reading context.source. If you do want to use it, you might need to merge main since I just did a bug fix for it

File: server/game/cards/01_SOR/units/EzraBridger.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+                    targetResolver: {
+                        mode: TargetMode.Select,
+                        choices: {
+                            ['Play it']: AbilityHelper.immediateEffects.playCardFromOutOfPlay(() => ({ target: topCardOfDeck })),
+                            ['Discard it']: AbilityHelper.immediateEffects.discardSpecificCard(() => ({ target: topCardOfDeck })),
+                            ['Leave it on top of your deck']: AbilityHelper.immediateEffects.noAction,
```
Comment (AMMayberry1) [inline server/game/cards/01_SOR/units/EzraBridger.ts:33]: I think there needs to be a parenthesis on the noAction like `AbilityHelper.immediateEffects.noAction()`. If you hit an issue where that makes this button disappear, then try `immediateEffects.noAction({ hasLegalTarget: true })`

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+            });
+
+            it('should be able to play it from the discard pile when an opponent\'s unit is defeated.', function () {
+                const { context } = contextRef;
+
+                context.player1.clickCard(context.vigilantHonorGuards);
+                context.player1.clickCard(context.wampa);
+
+                expect(context.player2).toBeActivePlayer();
+                expect(context.wampa).toBeInLocation('discard');
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:28]: This check isn't really needed

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+                context.player2.clickCard(context.confiscate);
+                expect(context.brutalTraditions).toBeInLocation('discard');
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:31]: Any reason to do it this way instead of starting it in discard? Could just use like atst instead of vigilant honor guards to kill wampa

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+                expect(context.player1).toBeActivePlayer();
+                expect(context.player1).toBeAbleToSelect(context.brutalTraditions);
+
+                context.player1.clickCard(context.brutalTraditions);
+                context.player1.clickCard(context.vigilantHonorGuards);
+                expect(context.vigilantHonorGuards.upgrades).toEqual([context.brutalTraditions]);
+                expect(context.vigilantHonorGuards.getPower()).toBe(5);
+                expect(context.vigilantHonorGuards.getHp()).toBe(8);
+                expect(context.vigilantHonorGuards.damage).toBe(4);
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:40]: Can remove these checks, we have other tests confirming that attached upgrades give stats correctly. Just checking the upgrade is attached is enough

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+                // Brutal tradition is again able to be played from the discard pile
+                context.player2.clickCard(context.vanquish);
+                expect(context.player2).toBeAbleToSelect(context.vigilantHonorGuards);
+                context.player2.clickCard(context.vigilantHonorGuards);
+                expect(context.vigilantHonorGuards).toBeInLocation('discard');
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:46]: Don't need this check, we know Vanquish works

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+                });
+            });
+
+            it('should be able to play it from the discard pile when an opponent\'s unit is defeated.', function () {
+                const { context } = contextRef;
+
+                context.player1.clickCard(context.atst);
+                context.player1.clickCard(context.wampa);
+                context.player2.passAction();
+                expect(context.player1).toBeAbleToSelect(context.brutalTraditions);
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:28]: One last check we can do I realized - now that we have your feature for highlighting selectable cards in the action phase, can you add a check that it correctly appears in the highlight list in the action phase? Same for other places we do this check in here

File: server/game/cards/01_SOR/units/EzraBridger.ts [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    public override setupCardAbilities() {
+        this.addTriggeredAbility({
+            title: 'Look at the top card of your deck.',
+            when: {
+                onAttackCompleted: (event, context) => event.attack.attacker === context.source,
+            },
+            immediateEffect: AbilityHelper.immediateEffects.lookAt(
+                (context) => ({ target: context.source.controller.getTopCardOfDeck() })
+            ),
+            ifYouDo: (thenContext) => {
```
Comment (AMMayberry1) [inline server/game/cards/01_SOR/units/EzraBridger.ts:23]: Do we actually need this context? The top card of deck is accessible via the regular context, so we could just do:

```typescript
ifYouDo: {
    title: 'You may play it, discard it, or leave it on top of your deck.',
    targetResolver: {
        mode: TargetMode.Select,
        choices: {
            ['Play it']: AbilityHelper.immediateEffects.playCardFromOutOfPlay((context) => ({ target: context.source.controller.getTopCardOfDeck() }))),
            ['Discard it']: AbilityHelper.immediateEffects.discardSpecificCard((context) => ({ target: context.source.controller.getTopCardOfDeck() }))),
            ['Leave it on top of your deck']: AbilityHelper.immediateEffects.noAction({ hasLegalTarget: true }),
        }
    }
}
```
If you do decide you prefer it the way it is, let's rename `thenContext` -> `ifYouDoContext`

Comment (CheBato) [reply server/game/cards/01_SOR/units/EzraBridger.ts:23]: you are right it is much cleaner this way. I changed it.

**Review by AMMayberry1 — APPROVED** [review]
